lib/opFns.js

```diff
@@ -418,25 +421,37 @@ module.exports = {
     })
   },
   JUMP: function (dest, runState) {
-    dest = utils.bufferToInt(dest)
+    dest = new BN(dest)
+
+    if (dest.gtn(runState.code.length)) {
+      trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
+    }
+
+    dest = dest.toNumber()
 
     if (!jumpIsValid(runState, dest)) {
       trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
     }
 
     runState.programCounter = dest
   },
-  JUMPI: function (c, i, runState) {
-    c = utils.bufferToInt(c)
-    i = utils.bufferToInt(i)
+  JUMPI: function (dest, cond, runState) {
+    dest = new BN(dest)
+    cond = new BN(cond)
 
-    var dest = i ? c : runState.programCounter
+    if (!cond.isZero()) {
+      if (dest.gtn(runState.code.length)) {
+        trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
+      }
 
-    if (i && !jumpIsValid(runState, dest)) {
-      trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
-    }
+      dest = dest.toNumber()
 
-    runState.programCounter = dest
+      if (!jumpIsValid(runState, dest)) {
+        trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
+      }
+
+      runState.programCounter = dest
+    }
   },
```

Comment:
Hm, that is the edge case of having `JUMPI` as the last instruction in the code. Need to check the YP.